store/tikv/coprocessor.go

```diff
@@ -219,6 +213,48 @@ const rangesPerTask = 25000
 
 func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv.Request) ([]*copTask, error) {
 	start := time.Now()
+	tasksCh := make(chan *copTask)
+
+	errCh, err := buildCopTasksChan(bo, cache, ranges, req, tasksCh, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var tasks []*copTask
+	for task := range tasksCh {
+		tasks = append(tasks, task)
+		select {
+		case err = <-errCh:
+			if err != nil {
+				return nil, err
+			}
+		default:
+		}
+	}
+
+	if elapsed := time.Since(start); elapsed > time.Millisecond*500 {
+		logutil.BgLogger().Warn("buildCopTasks takes too much time",
+			zap.Duration("elapsed", elapsed),
+			zap.Int("range len", ranges.len()),
+			zap.Int("task len", len(tasks)))
+	}
+	tikvTxnRegionsNumHistogramWithCoprocessor.Observe(float64(len(tasks)))
+	return tasks, nil
+}
+
+// shouldStop reads finishCh and returns a boolean value about should we stop to do our work or not
+func shouldStop(finishCh <-chan struct{}) bool {
+	select {
+	case <-finishCh:
+		return true
+	default:
+		return false
+	}
+}
+
+// buildCopTasksChan returns a channel to get tasks
+func buildCopTasksChan(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv.Request, tasksCh chan *copTask, finishCh <-chan struct{}) (errCh chan error, err error) {
+	errCh = make(chan error)
```

Comment:
Best to make 1 channel buffer to avoid the caller does not receive the `errCh` immediately.

Reply:
Done